Index: src/vobject/icalendar.py
===================================================================
--- src/vobject/icalendar.py	(revision 193)
+++ src/vobject/icalendar.py	(working copy)
@@ -1601,7 +1601,7 @@
     charIterator = enumerate(s)
     state        = "read normal"
 
-    current = ""
+    current = []
     results = []
 
     while True:
@@ -1615,28 +1615,31 @@
                 state = "read escaped char"
             elif char == listSeparator:
                 state = "read normal"
+                current = "".join(current)
                 results.append(current)
-                current = ""
+                current = []
             elif char == "eof":
                 state = "end"
             else:
                 state = "read normal"
-                current = current + char
+                current.append(char)
 
         elif state == "read escaped char":
             if escapableChar(char):
                 state = "read normal"
                 if char in 'nN': 
-                    current = current + '\n'
+                    current.append('\n')
                 else:
-                    current = current + char
+                    current.append(char)
             else:
-                state = "read normal"
+                #state = "read normal"
                 # leave unrecognized escaped characters for later passes
-                current = current + '\\' + char 
+                #current.append('\\' + char)
+                raise ParseError("error: illegal escape sequence: '\\%s'" % (char,))
 
         elif state == "end":    #an end state
-            if current != "" or len(results) == 0:
+            if len(current) or len(results) == 0:
+                current = "".join(current)
                 results.append(current)
             return results
 
