Index: vobject/vcard.py
===================================================================
--- vobject/vcard.py	(revision 218)
+++ vobject/vcard.py	(working copy)
@@ -196,7 +196,7 @@
            can handle PHOTO if it's not wrapped, so don't wrap it."""
         if wacky_apple_photo_serialize:
             lineLength = REALLY_LARGE
-        VCardTextBehavior.serialize(obj, buf, lineLength, validate)
+        return VCardTextBehavior.serialize(obj, buf, lineLength, validate)
 
 registerBehavior(Photo)
 
@@ -207,10 +207,10 @@
     else:
         return stringList
 
-def splitFields(string):
+def splitFields(string, maybeList=True):
     """Return a list of strings or lists from a Name or Address."""
-    return [toListOrString(i) for i in
-            stringToTextValues(string, listSeparator=';', charList=';')]
+    return [(toListOrString(i) if maybeList else i) for i in
+            stringToTextValues(string, listSeparator=';')]
 
 def toList(stringOrList):
     if isinstance(stringOrList, basestring):
@@ -287,7 +287,7 @@
         """Turn obj.value into a list."""
         if obj.isNative: return obj
         obj.isNative = True
-        obj.value = splitFields(obj.value)
+        obj.value = splitFields(obj.value, maybeList=False)
         return obj
 
     @staticmethod
