Index: vobject/icalendar.py
===================================================================
--- vobject/icalendar.py	(revision 219)
+++ vobject/icalendar.py	(working copy)
@@ -1675,9 +1675,10 @@
                 else:
                     current.append(char)
             else:
-                state = "read normal"
+                #state = "read normal"
                 # leave unrecognized escaped characters for later passes
-                current.append('\\' + char)
+                #current.append('\\' + char)
+                raise ParseError("error: illegal escape sequence: '\\%s'" % (char,))
 
         elif state == "end":    #an end state
             if len(current) or len(results) == 0:
