Index: vobject/base.py
===================================================================
--- vobject/base.py	(revision 219)
+++ vobject/base.py	(working copy)
@@ -1051,10 +1051,11 @@
                         component=stack.pop()
                         if versionLine is not None:
                             component.setBehaviorFromVersionLine(versionLine)
-                        else:
-                            behavior = getBehavior(component.name)
-                            if behavior:
-                                component.setBehavior(behavior)
+                        # calendarserver.org: be strict about what we accept
+                        #else:
+                        #    behavior = getBehavior(component.name)
+                        #    if behavior:
+                        #        component.setBehavior(behavior)
                         if validate: component.validate(raiseException=True)
                         if transform: component.transformChildrenToNative()
                         yield component #EXIT POINT
