diff --git a/twistedcaldav/authkerb.py b/twistedcaldav/authkerb.py
index 4f9abe3..6cdb7f0 100644
--- a/twistedcaldav/authkerb.py
+++ b/twistedcaldav/authkerb.py
@@ -84,12 +84,13 @@ class KerberosCredentialFactoryBase(object):
             splits = principal.split("/")
             servicetype = splits[0]
             splits = splits[1].split("@")
+            server = splits[0]
             realm = splits[1]
         except IndexError:
             logging.err("Invalid Kerberos principal: %s" % (principal,), system="KerberosCredentialFactoryBase")
             raise ValueError('Authentication System Failure: Invalid Kerberos principal: %s' % (principal,))
                 
-        self.service = "%s@%s" % (servicetype, realm,)
+        self.service = "%s@%s" % (servicetype, server,)
         self.realm = realm
 
 class BasicKerberosCredentials(credentials.UsernamePassword):
