diff --git a/pysrc/kerberos.py b/pysrc/kerberos.py
index f40f4e9..fae50e7 100644
--- a/pysrc/kerberos.py
+++ b/pysrc/kerberos.py
@@ -55,6 +55,19 @@ def checkPassword(user, pswd, service, default_realm):
     @return:              True if authentication succeeds, False otherwise.
     """
 
+def changePassword(user, oldpswd, newpswd):
+    """
+    This function allows to change the user password on the KDC.
+
+    @param user:          a string containing the Kerberos user name. A realm may be
+        included by appending an '@' followed by the realm string to the actual user id.
+        If no realm is supplied, then the realm set in the default_realm argument will
+        be used.
+    @param oldpswd:       a string containing the old (current) password for the user.
+    @param newpswd:       a string containging the new password for the user.
+    @return:              True if password changing succeeds, False otherwise.
+    """
+
 def getServerPrincipalDetails(service, hostname):
     """
     This function returns the service principal for the server given a service type
