Index: twisted/web2/dav/test/test_report.py
===================================================================
--- twisted/web2/dav/test/test_report.py	(revision 27622)
+++ twisted/web2/dav/test/test_report.py	(working copy)
@@ -34,6 +34,16 @@
     """
     REPORT request
     """
+
+    def setUp(self):
+        # Pre-fetch site, so the rest of the test doesn't have to defer
+        twisted.web2.dav.test.util.TestCase.setUp(self)
+        self._getSite()
+
+    def _getStoredSite(self):
+        return self._site
+    site = property(_getStoredSite)
+
     def test_REPORT_no_body(self):
         """
         REPORT request with no body
