Index: twisted/python/versions.py
===================================================================
--- twisted/python/versions.py	(revision 18219)
+++ twisted/python/versions.py	(working copy)
@@ -91,7 +91,7 @@
         mod = sys.modules.get(self.package)
         if mod:
             rev = os.popen("svnversion %r" % (os.path.dirname(mod.__file__),)).read()
-            return rev
+            return rev.replace("\n", "")
 
     def _formatSVNVersion(self):
         ver = self._getSVNVersion()
