Index: twisted/spread/pb.py
===================================================================
--- twisted/spread/pb.py	(revision 19773)
+++ twisted/spread/pb.py	(working copy)
@@ -64,7 +64,11 @@
 except ImportError:
     import StringIO
 
-import md5
+try:
+    from hashlib import md5
+except ImportError:
+    import md5.new as md5
+
 import random
 import new
 import types
@@ -1003,10 +1007,10 @@
 
     This is useful for challenge/response authentication.
     """
-    m = md5.new()
+    m = md5()
     m.update(password)
     hashedPassword = m.digest()
-    m = md5.new()
+    m = md5()
     m.update(hashedPassword)
     m.update(challenge)
     doubleHashedPassword = m.digest()
@@ -1017,7 +1021,7 @@
     crap = ''
     for x in range(random.randrange(15,25)):
         crap = crap + chr(random.randint(65,90))
-    crap = md5.new(crap).digest()
+    crap = md5(crap).digest()
     return crap
 
 
@@ -1226,11 +1230,11 @@
 
     # IUsernameHashedPassword:
     def checkPassword(self, password):
-        return self.checkMD5Password(md5.md5(password).digest())
+        return self.checkMD5Password(md5(password).digest())
 
     # IUsernameMD5Password
     def checkMD5Password(self, md5Password):
-        md = md5.new()
+        md = md5()
         md.update(md5Password)
         md.update(self.challenge)
         correct = md.digest()
