Index: twisted/web2/dav/resource.py
===================================================================
--- twisted/web2/dav/resource.py	(revision 26969)
+++ twisted/web2/dav/resource.py	(working copy)
@@ -49,12 +49,13 @@
 if not hasattr(__builtin__, "frozenset"):
     import sets.ImmutableSet as frozenset
 
+import urllib
+
 from zope.interface import implements
 from twisted.python import log
 from twisted.python.failure import Failure
 from twisted.internet.defer import Deferred, maybeDeferred, succeed
 from twisted.internet.defer import waitForDeferred, deferredGenerator
-from twisted.cred.error import LoginFailed, UnauthorizedLogin
 from twisted.internet import reactor
 from twisted.web2 import responsecode
 from twisted.web2.http import HTTPError, RedirectResponse, StatusResponse
@@ -1880,7 +1881,7 @@
         # If this is a collection and the URI doesn't end in "/", redirect.
         #
         if self.isCollection() and request.path[-1:] != "/":
-            return RedirectResponse(request.unparseURL(path=request.path+'/'))
+            return RedirectResponse(request.unparseURL(path=urllib.quote(urllib.unquote(request.path), safe=':/')+'/'))
 
         def setHeaders(response):
             response = IResponse(response)
