Index: twisted/web2/dav/resource.py
===================================================================
--- twisted/web2/dav/resource.py	(revision 27622)
+++ twisted/web2/dav/resource.py	(working copy)
@@ -706,7 +706,7 @@
                 hasattr(request, 'loginInterfaces')):
             request.authnUser = davxml.Principal(davxml.Unauthenticated())
             request.authzUser = davxml.Principal(davxml.Unauthenticated())
-            return (request.authnUser, request.authzUser)
+            return succeed((request.authnUser, request.authzUser))
 
         authHeader = request.headers.getHeader('authorization')
 
@@ -754,7 +754,7 @@
         else:
             request.authnUser = davxml.Principal(davxml.Unauthenticated())
             request.authzUser = davxml.Principal(davxml.Unauthenticated())
-            return (request.authnUser, request.authzUser)
+            return succeed((request.authnUser, request.authzUser))
 
     ##
     # ACL
