Index: twisted/web2/dav/noneprops.py
===================================================================
--- twisted/web2/dav/noneprops.py	(revision 27622)
+++ twisted/web2/dav/noneprops.py	(working copy)
@@ -33,6 +33,8 @@
 
 from twisted.web2 import responsecode
 from twisted.web2.http import HTTPError, StatusResponse
+from twisted.python.failure import Failure
+from twisted.internet.defer import succeed, fail
 
 class NonePropertyStore (object):
     """
@@ -50,18 +52,18 @@
         pass
 
     def get(self, qname):
-        raise HTTPError(StatusResponse(responsecode.NOT_FOUND, "No such property: {%s}%s" % qname))
+        return fail(Failure(HTTPError(StatusResponse(responsecode.NOT_FOUND, "No such property: {%s}%s" % qname))))
 
     def set(self, property):
-        raise HTTPError(StatusResponse(responsecode.FORBIDDEN, "Permission denied for setting property: %s" % (property,)))
+        return fail(Failure(HTTPError(StatusResponse(responsecode.FORBIDDEN, "Permission denied for setting property: %s" % (property,)))))
 
     def delete(self, qname):
         # RFC 2518 Section 12.13.1 says that removal of
         # non-existing property is not an error.
-        pass
+        return succeed(None)
 
     def contains(self, qname):
-        return False
+        return succeed(False)
 
     def list(self):
-        return ()
+        return succeed( () )
