Index: twisted/python/procutils.py
===================================================================
--- twisted/python/procutils.py	(revision 18545)
+++ twisted/python/procutils.py	(working copy)
@@ -33,7 +33,7 @@
     """
     result = []
     exts = filter(None, os.environ.get('PATHEXT', '').split(os.pathsep))
-    for p in os.environ['PATH'].split(os.pathsep):
+    for p in os.environ.get('PATH', '').split(os.pathsep):
         p = os.path.join(p, name)
         if os.access(p, flags):
             result.append(p)
