Index: twisted/plugins/__init__.py
===================================================================
--- twisted/plugins/__init__.py	(revision 19773)
+++ twisted/plugins/__init__.py	(working copy)
@@ -12,6 +12,6 @@
 """
 
 import os, sys
-__path__ = [os.path.abspath(os.path.join(x, 'twisted', 'plugins')) for x in sys.path]
+__path__ = [os.path.abspath(os.path.join(x, 'twisted', 'plugins')) for x in sys.path if not x.startswith('/System')]
 
 __all__ = []                    # nothing to see here, move along, move along
