Index: twisted/web2/dav/element/parser.py
===================================================================
--- twisted/web2/dav/element/parser.py	(revision 27622)
+++ twisted/web2/dav/element/parser.py	(working copy)
@@ -54,7 +54,8 @@
     """
     element_names = []
 
-    for element_class_name in dir(module):
+    items = module.__all__ if hasattr(module, "__all__") else dir(module)
+    for element_class_name in items:
         element_class = getattr(module, element_class_name)
 
         if type(element_class) is type and issubclass(element_class, WebDAVElement):
