Index: twisted/web2/dav/static.py
===================================================================
--- twisted/web2/dav/static.py	(revision 18074)
+++ twisted/web2/dav/static.py	(working copy)
@@ -34,6 +34,8 @@
 from twisted.internet.defer import succeed, deferredGenerator, waitForDeferred
 from twisted.web2.static import File
 from twisted.web2 import dirlist
+from twisted.web2 import http
+from twisted.web2 import responsecode
 from twisted.web2.dav import davxml
 from twisted.web2.dav.idav import IDAVResource
 from twisted.web2.dav.resource import DAVResource
