Index: twisted/web2/filter/location.py
===================================================================
--- twisted/web2/filter/location.py	(revision 18208)
+++ twisted/web2/filter/location.py	(working copy)
@@ -18,10 +18,12 @@
             #
             (scheme, host, path, params, querystring, fragment) = urlparse.urlparse(location)
 
-            if scheme == "":
-                location = request.unparseURL(path=location)
+            if scheme == '':
+                abslocation = request.unparseURL(path=location)
+            else:
+                abslocation = location
         
-            response.headers.setHeader("location", location)
+            response.headers.setHeader("location", abslocation)
 
         return response
 
