Index: twisted/internet/defer.py
===================================================================
--- twisted/internet/defer.py	(revision 19773)
+++ twisted/internet/defer.py	(working copy)
@@ -998,6 +998,6 @@
 __all__ = ["Deferred", "DeferredList", "succeed", "fail", "FAILURE", "SUCCESS",
            "AlreadyCalledError", "TimeoutError", "gatherResults",
            "maybeDeferred",
-           "waitForDeferred", "deferredGenerator", "inlineCallbacks",
+           "waitForDeferred", "deferredGenerator", "returnValue", "inlineCallbacks",
            "DeferredLock", "DeferredSemaphore", "DeferredQueue",
           ]
