=== modified file 'twisted/web2/dav/static.py'
--- twisted/web2/dav/static.py	2009-04-20 17:38:06 +0000
+++ twisted/web2/dav/static.py	2009-07-15 23:30:18 +0000
@@ -53,6 +53,14 @@
 
     Extends twisted.web2.static.File to handle WebDAV methods.
     """
+    @classmethod
+    def fetch(cls, request, path, *args, **kwargs):
+        """
+        stuff etc
+        """
+        return succeed(cls(path, *args, **kwargs))
+
+
     def __init__(
         self, path,
         defaultType="text/plain", indexNames=None,
@@ -192,7 +200,7 @@
         return (self.createSimilarFile(self.fp.child(path).path), segments[1:])
 
     def createSimilarFile(self, path):
-        return self.__class__(
+        return self.__class__.fetch(None, 
             path, defaultType=self.defaultType, indexNames=self.indexNames[:],
             principalCollections=self.principalCollections())
 

