Index: twisted/conch/test/test_keys.py
===================================================================
--- twisted/conch/test/test_keys.py	(revision 19773)
+++ twisted/conch/test/test_keys.py	(working copy)
@@ -4,10 +4,10 @@
 
 try:
     import Crypto
+    from twisted.conch.ssh import keys
 except ImportError:
     Crypto = None
 
-from twisted.conch.ssh import keys
 from twisted.trial import unittest
 
 publicRSA_openssh = "ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAGEArzJx8OYOnJmzf4tfBEvLi8DVPrJ3/c9k2I/Az64fxjHf9imyRJbixtQhlH9lfNjUIx+4LmrJH5QNRsFporcHDKOTwTTYLh5KmRpslkYHRivcJSkbh/C+BR3utDS555mV comment"
