Index: twisted/web2/dav/fileop.py
===================================================================
--- twisted/web2/dav/fileop.py	(revision 19773)
+++ twisted/web2/dav/fileop.py	(working copy)
@@ -35,6 +35,7 @@
     "move",
     "put",
     "mkcollection",
+    "rmdir",
 ]
 
 import os
@@ -287,7 +288,7 @@
                     response = waitForDeferred(copy(FilePath(source_path), FilePath(destination_path), destination_uri, depth))
                     yield response
                     response = response.getResult()
-                    checkResponse(response, "copy", responsecode.NO_CONTENT)
+                    checkResponse(response, "copy", responsecode.CREATED, responsecode.NO_CONTENT)
 
             for subdir in subdirs:
                 source_path, destination_path = paths(dir, subdir)
@@ -509,7 +510,5 @@
     os.rmdir(dirname)
 
 def checkResponse(response, method, *codes):
-    assert (
-        response in codes,
-        "%s() should have raised, but returned one of %r instead" % (method, codes)
-    )
+    assert  response in codes, \
+        "%s() returned %r, but should have returned one of %r instead" % (method, response, codes)
