Index: twisted/web2/auth/digest.py
===================================================================
--- twisted/web2/auth/digest.py	(revision 26741)
+++ twisted/web2/auth/digest.py	(working copy)
@@ -158,6 +158,17 @@
             algo, nonce, nc, cnonce, qop, self.method, uri, None
         )
 
+        if expected == response:
+            return True
+
+        # IE7 sends cnonce and nc values, but auth fails if they are used.
+        # So try again without them...
+        # They can be omitted for backwards compatibility [RFC 2069].
+        expected = calcResponse(
+            calcHA1(algo, self.username, self.realm, password, nonce, cnonce),
+            algo, nonce, None, None, qop, self.method, uri, None
+        )
+
         return expected == response
 
     def checkHash(self, digestHash):
