Index: twisted/web2/dav/test/test_copy.py
===================================================================
--- twisted/web2/dav/test/test_copy.py	(revision 19773)
+++ twisted/web2/dav/test/test_copy.py	(working copy)
@@ -22,9 +22,9 @@
 # DRI: Wilfredo Sanchez, wsanchez@apple.com
 ##
 
+from hashlib import md5
 import os
 import urllib
-import md5
 
 import twisted.web2.dav.test.util
 from twisted.web2 import responsecode
@@ -161,7 +161,7 @@
             yield (request, do_test)
 
 def sumFile(path):
-    m = md5.new()
+    m = md5()
 
     if os.path.isfile(path):
         f = file(path)
