Index: twisted/web2/dav/util.py
===================================================================
--- twisted/web2/dav/util.py	(revision 19773)
+++ twisted/web2/dav/util.py	(working copy)
@@ -37,6 +37,7 @@
     "normalizeURL",
     "joinURL",
     "parentForURL",
+    "unimplemented",
     "bindMethods",
 ]
 
@@ -76,7 +77,9 @@
 
     def parse(xml):
         try:
-            return davxml.WebDAVDocument.fromString(xml)
+            doc = davxml.WebDAVDocument.fromString(xml)
+            doc.root_element.validate()
+            return doc
         except ValueError:
             log.err("Bad XML:\n%s" % (xml,))
             raise
