Index: twisted/mail/imap4.py
===================================================================
--- twisted/mail/imap4.py	(revision 27622)
+++ twisted/mail/imap4.py	(working copy)
@@ -391,10 +391,10 @@
         return ''
 
     def setResponse(self, response):
-        parts = response[:-1].split('\0', 1)
-        if len(parts) != 2:
+        parts = response.split('\0', 2)
+        if len(parts) != 3:
             raise IllegalClientResponse("Malformed Response - wrong number of parts")
-        self.username, self.password = parts
+        _ignore, self.username, self.password = parts
 
     def moreChallenges(self):
         return False
