Index: twisted/web2/dav/method/report.py
===================================================================
--- twisted/web2/dav/method/report.py	(revision 19773)
+++ twisted/web2/dav/method/report.py	(working copy)
@@ -27,7 +27,11 @@
 WebDAV REPORT method
 """
 
-__all__ = ["http_REPORT"]
+__all__ = [
+    "max_number_of_matches",
+    "NumberOfMatchesWithinLimits",
+    "http_REPORT",
+]
 
 import string
 
@@ -43,7 +47,14 @@
 max_number_of_matches = 500
 
 class NumberOfMatchesWithinLimits(Exception):
-    pass
+    
+    def __init__(self, limit):
+        
+        super(NumberOfMatchesWithinLimits, self).__init__()
+        self.limit = limit
+        
+    def maxLimit(self):
+        return self.limit
 
 def http_REPORT(self, request):
     """
