Index: twisted/python/failure.py
===================================================================
--- twisted/python/failure.py	(revision 19773)
+++ twisted/python/failure.py	(working copy)
@@ -325,7 +325,8 @@
         ]
 
         # added 2003-06-23. See comment above in __init__
-        c['tb'] = None
+        # Actually we need the self.tb for when we raise using self.raiseException
+        #c['tb'] = None
 
         if self.stack is not None:
             # XXX: This is a band-aid.  I can't figure out where these
