Index: twisted/test/test_plugin.py
===================================================================
--- twisted/test/test_plugin.py	(revision 19773)
+++ twisted/test/test_plugin.py	(working copy)
@@ -518,3 +518,6 @@
         self.assertEqual(len(self.flushLoggedErrors()), 0)
         self.assertIn('one', self.getAllPlugins())
         self.assertEqual(len(self.flushLoggedErrors()), 1)
+
+    
+    test_newPluginsOnReadOnlyPath.skip = "Seems not to work on OS X 10.4 buildbot machine."
