Index: twisted/web2/dav/test/util.py
===================================================================
--- twisted/web2/dav/test/util.py	(revision 19773)
+++ twisted/web2/dav/test/util.py	(working copy)
@@ -184,10 +184,11 @@
         d.addCallback(lambda resource: resource.renderHTTP(request))
         d.addCallback(request._cbFinishRender)
 
-        if type(callback) is tuple:
-            d.addCallbacks(*callback)
-        else:
-            d.addCallback(callback)
+        if callback:
+            if type(callback) is tuple:
+                d.addCallbacks(*callback)
+            else:
+                d.addCallback(callback)
 
         return d
 
