=== modified file 'twisted/application/app.py'
--- twisted/application/app.py	2008-12-14 22:09:07 +0000
+++ twisted/application/app.py	2009-07-15 23:30:17 +0000
@@ -241,7 +241,11 @@
         @param application: The application on which to check for an
             L{ILogObserver}.
         """
-        observer = application.getComponent(ILogObserver, None)
+        
+        if application:
+            observer = application.getComponent(ILogObserver, None)
+        else:
+            observer = None
 
         if observer is None:
             observer = self._getLogObserver()
@@ -408,8 +412,6 @@
         Run the application.
         """
         self.preApplication()
-        self.application = self.createOrGetApplication()
-
 
         getLogObserverLegacy = getattr(self, 'getLogObserver', None)
         if getLogObserverLegacy is not None:
@@ -418,7 +420,9 @@
                           category=DeprecationWarning)
             self.startLogging(self.getLogObserver())
         else:
-            self.logger.start(self.application)
+            self.logger.start(None)
+
+        self.application = self.createOrGetApplication()
 
         self.postApplication()
         self.logger.stop()

