Index: twisted/web2/dav/element/rfc2518.py
===================================================================
--- twisted/web2/dav/element/rfc2518.py	(revision 19773)
+++ twisted/web2/dav/element/rfc2518.py	(working copy)
@@ -59,8 +59,8 @@
     """
     name = "depth"
 
-    def __init__(self, *children, **attributes):
-        super(Depth, self).__init__(*children, **attributes)
+    def validate(self):
+        super(Depth, self).validate()
 
         depth = str(self)
         if depth not in ("0", "1", "infinity"):
@@ -382,8 +382,8 @@
         PCDATAElement: (0, 1),
     }
 
-    def __init__(self, *children, **attributes):
-        super(KeepAlive, self).__init__(*children, **attributes)
+    def validate(self):
+        super(KeepAlive, self).validate()
 
         type = None
 
@@ -450,8 +450,8 @@
         (dav_namespace, "prop"    ): (0, 1),
     }
 
-    def __init__(self, *children, **attributes):
-        super(PropertyFind, self).__init__(*children, **attributes)
+    def validate(self):
+        super(PropertyFind, self).validate()
 
         if len(self.children) != 1:
             raise ValueError(
