=== modified file 'twisted/web2/error.py'
--- twisted/web2/error.py	2005-11-22 22:59:59 +0000
+++ twisted/web2/error.py	2009-07-15 23:30:18 +0000
@@ -92,7 +92,7 @@
             "<body><h1>%s</h1>%s</body></html>") % (
         response.code, title, title, message)
     
-    response.headers.setHeader("content-type", http_headers.MimeType('text', 'html'))
+    response.headers.setHeader("content-type", http_headers.MimeType('text', 'html', {'charset':'utf-8'}))
     response.stream = stream.MemoryStream(body)
     
     return response

