Index: twisted/web2/dav/davxml.py
===================================================================
--- twisted/web2/dav/davxml.py	(revision 19773)
+++ twisted/web2/dav/davxml.py	(working copy)
@@ -45,6 +45,8 @@
 from twisted.web2.dav.element.rfc2518 import *
 from twisted.web2.dav.element.rfc3253 import *
 from twisted.web2.dav.element.rfc3744 import *
+from twisted.web2.dav.element.rfc4331 import *
+from twisted.web2.dav.element.extensions import *
 
 #
 # Register all XML elements with the parser
@@ -56,11 +58,15 @@
 import twisted.web2.dav.element.rfc2518
 import twisted.web2.dav.element.rfc3253
 import twisted.web2.dav.element.rfc3744
+import twisted.web2.dav.element.rfc4331
+import twisted.web2.dav.element.extensions
 
 __all__ = (
     registerElements(twisted.web2.dav.element.base   ) +
     registerElements(twisted.web2.dav.element.parser ) +
     registerElements(twisted.web2.dav.element.rfc2518) +
     registerElements(twisted.web2.dav.element.rfc3253) +
-    registerElements(twisted.web2.dav.element.rfc3744)
+    registerElements(twisted.web2.dav.element.rfc3744) +
+    registerElements(twisted.web2.dav.element.rfc4331) +
+    registerElements(twisted.web2.dav.element.extensions)
 )
