Index: twisted/web2/dav/resource.py
===================================================================
--- twisted/web2/dav/resource.py	(revision 18074)
+++ twisted/web2/dav/resource.py	(working copy)
@@ -599,7 +599,7 @@
             else:
                 factory = request.credentialFactories[authHeader[0]]
 
-                creds = factory.decode(authHeader[1], request.method)
+                creds = factory.decode(authHeader[1], request)
 
                 # Try to match principals in each principal collection on 
                 # the resource
