Index: twisted/web2/dav/http.py
===================================================================
--- twisted/web2/dav/http.py	(revision 17005)
+++ twisted/web2/dav/http.py	(working copy)
@@ -102,7 +102,8 @@
             else:
                 uri = joinURL(base_uri, subpath)
 
-            denials.append(davxml.Resource(davxml.HRef(uri), *[davxml.Privilege(p) for p in privileges]))
+            for p in privileges:
+                denials.append(davxml.Resource(davxml.HRef(uri), davxml.Privilege(p)))
 
         super(NeedPrivilegesResponse, self).__init__(responsecode.FORBIDDEN, davxml.NeedPrivileges(*denials))
 
@@ -115,9 +116,6 @@
         """
         @param xml_responses: an interable of davxml.Response objects.
         """
-        multistatus = davxml.MultiStatus(*xml_responses)
-        output = multistatus.toxml()
-
         Response.__init__(self, code=responsecode.MULTI_STATUS,
                           stream=davxml.MultiStatus(*xml_responses).toxml())
 
