Index: twisted/web2/dav/acl.py
===================================================================
--- twisted/web2/dav/acl.py	(revision 17005)
+++ twisted/web2/dav/acl.py	(working copy)
@@ -50,8 +50,8 @@
     ##
 
     liveProperties = DAVLeafResource.liveProperties + (
-        (dav_namespace, "alternate-uri-set"),
-        (dav_namespace, "principal-url"    ),
+        (dav_namespace, "alternate-URI-set"),
+        (dav_namespace, "principal-URL"    ),
         (dav_namespace, "group-member-set" ),
         (dav_namespace, "group-membership" ),
     )
@@ -68,10 +68,8 @@
     def readProperty(self, property, request):
         if type(property) is tuple:
             qname = property
-            sname = "{%s}%s" % property
         else:
             qname = property.qname()
-            sname = property.sname()
 
         namespace, name = qname
 
@@ -76,10 +74,10 @@
         namespace, name = qname
 
         if namespace == dav_namespace:
-            if name == "alternate-uri-set":
+            if name == "alternate-URI-set":
                 return davxml.AlternateURISet(*[davxml.HRef(u) for u in self.alternateURIs()])
 
-            if name == "principal-url":
+            if name == "principal-URL":
                 return davxml.PrincipalURL(davxml.HRef(self.principalURL()))
 
             if name == "group-member-set":
@@ -86,7 +84,7 @@
                 return davxml.GroupMemberSet(*[davxml.HRef(p) for p in self.groupMembers()])
 
             if name == "group-membership":
-                return davxml.GroupMemberSet(*[davxml.HRef(g) for g in self.groupMemberships()])
+                return davxml.GroupMembership(*[davxml.HRef(g) for g in self.groupMemberships()])
 
             if name == "resourcetype":
                 if self.isCollection():
