Index: twisted/web2/dav/auth.py
===================================================================
--- twisted/web2/dav/auth.py	(revision 17966)
+++ twisted/web2/dav/auth.py	(working copy)
@@ -8,8 +8,6 @@
 
 __all__ = ["PrincipalCredentials", "AuthenticationWrapper"]
 
-__all__ = ["PrincipalCredentials"]
-
 class AuthenticationWrapper(WrapperResource):
     def __init__(self, resource, portal, credentialFactories, loginInterfaces):
         """Wrap the given resource and use the parameters to set up the request
