Index: twisted/web2/dav/static.py
===================================================================
--- twisted/web2/dav/static.py	(revision 17940)
+++ twisted/web2/dav/static.py	(working copy)
@@ -139,7 +139,7 @@
                     ),
                 ),
             )
-        return DAVFile._supportedPrivilegeSet
+        return succeed(DAVFile._supportedPrivilegeSet)
 
     ##
     # Workarounds for issues with File
