Index: twisted/web2/auth/interfaces.py
===================================================================
--- twisted/web2/auth/interfaces.py	(revision 17932)
+++ twisted/web2/auth/interfaces.py	(working copy)
@@ -13,7 +13,7 @@
         @return: dictionary of challenge arguments
         """
 
-    def decode(response, method=None):
+    def decode(response, request):
         """Create a credentials object from the given response.
         May raise twisted.cred.error.LoginFailed if the response is invalid.
     
@@ -20,8 +20,8 @@
         @type response: C{str}
         @param response: scheme specific response string
 
-        @type method: C{str}
-        @param method: the method by which this response was sent
+        @type request: L{twisted.web2.server.Request}
+        @param request: the request being processed
 
         @return: ICredentials
         """
