Index: twisted/web2/dav/method/propfind.py
===================================================================
--- twisted/web2/dav/method/propfind.py	(revision 17951)
+++ twisted/web2/dav/method/propfind.py	(working copy)
@@ -48,6 +48,13 @@
         raise HTTPError(responsecode.NOT_FOUND)
 
     #
+    # Check authentication and access controls
+    #
+    x = waitForDeferred(self.securityCheck(request, (davxml.Read(),)))
+    yield x
+    x.getResult()
+
+    #
     # Read request body
     #
     try:
@@ -99,7 +106,7 @@
 
     resources = [(self, None)]
 
-    d = self.findChildren(depth, request, lambda x, y: resources.append((x, y)))
+    d = self.findChildren(depth, request, lambda x, y: resources.append((x, y)), (davxml.Read(),))
     ign = waitForDeferred(d)
     yield ign
     ign.getResult()
@@ -108,8 +115,6 @@
         if uri is None:
             uri = normalizeURL(request_uri)
             if self.isCollection() and not uri.endswith("/"): uri += "/"
-        else:
-            uri = joinURL(request_uri, uri)
 
         resource_properties = waitForDeferred(resource.listProperties(request))
         yield resource_properties
