Index: twisted/web2/dav/method/report_principal_match.py
===================================================================
--- twisted/web2/dav/method/report_principal_match.py	(revision 17951)
+++ twisted/web2/dav/method/report_principal_match.py	(working copy)
@@ -104,7 +104,7 @@
                     prop_common.responseForHref(
                         request,
                         responses,
-                        davxml.HRef.fromString(joinURL(request.uri, uri)),
+                        davxml.HRef.fromString(uri),
                         child,
                         propertiesForResource,
                         propElement
@@ -133,7 +133,7 @@
                             prop_common.responseForHref(
                                 request,
                                 responses,
-                                davxml.HRef.fromString(joinURL(request.uri, uri)),
+                                davxml.HRef.fromString(uri),
                                 child,
                                 propertiesForResource,
                                 propElement
