Index: run
===================================================================
--- run (revision 6280)
+++ run (working copy)
@@ -106,8 +106,6 @@
   fi;
 }
 
-. "${wd}/support/build.sh";
-
 # Actually run the server.  (Or, exit, if things aren't sufficiently set up in
 # order to do that.)
 run () {
@@ -149,6 +147,8 @@
 run_main () {
   parse_options "$@";
 
+  . "${wd}/support/build.sh";
+
   # If we've been asked to read a configuration key, just read it and exit.
   if [ -n "${read_key}" ]; then
     conf_read_key "${read_key}";
