Index: src/kerberos.c
===================================================================
--- src/kerberos.c	(revision 12223)
+++ src/kerberos.c	(working copy)
@@ -272,6 +272,30 @@
 	return Py_BuildValue("i", result);
 }
 
+static PyObject *authGSSClientInquireCred(PyObject *self, PyObject *args)
+{
+	gss_client_state *state;
+	PyObject *pystate;
+	int result = 0;
+	if (!PyArg_ParseTuple(args, "O", &pystate))
+		return NULL;
+
+	if (!PyCObject_Check(pystate)) {
+		PyErr_SetString(PyExc_TypeError, "Expected a context object");
+		return NULL;
+	}
+
+	state = (gss_client_state *)PyCObject_AsVoidPtr(pystate);
+	if (state == NULL)
+		return NULL;
+
+	result = authenticate_gss_client_inquire_cred(state);
+	if (result == AUTH_GSS_ERROR)
+		return NULL;
+
+	return Py_BuildValue("i", result);
+}
+
 static PyObject *authGSSServerInit(PyObject *self, PyObject *args)
 {
     const char *service = NULL;
@@ -423,6 +447,8 @@
      "return 1 if confidentiality was set in the last unwrapped buffer, 0 otherwise."},
     {"authGSSClientUserName",  authGSSClientUserName, METH_VARARGS,
      "Get the user name from the last client-side GSSAPI step."},
+    {"authGSSClientInquireCred",  authGSSClientInquireCred, METH_VARARGS,
+     "Get the current user name, if any, without a client-side GSSAPI step"},
     {"authGSSServerInit",  authGSSServerInit, METH_VARARGS,
      "Initialize server-side GSSAPI operations."},
     {"authGSSClientWrap",  authGSSClientWrap, METH_VARARGS,
@@ -429,6 +455,8 @@
      "Do a GSSAPI wrap."},
     {"authGSSClientUnwrap",  authGSSClientUnwrap, METH_VARARGS,
      "Do a GSSAPI unwrap."},
+    {"authGSSClientInquireCred", authGSSClientInquireCred, METH_VARARGS,
+     "Get the current user name, if any."},
     {"authGSSServerClean",  authGSSServerClean, METH_VARARGS,
      "Terminate server-side GSSAPI operations."},
     {"authGSSServerStep",  authGSSServerStep, METH_VARARGS,
Index: src/kerberosgss.c
===================================================================
--- src/kerberosgss.c	(revision 12223)
+++ src/kerberosgss.c	(working copy)
@@ -429,6 +429,68 @@
 	return ret;
 }
 
+int authenticate_gss_client_inquire_cred(gss_client_state* state)
+{
+    OM_uint32 maj_stat;
+    OM_uint32 min_stat;
+    gss_cred_id_t server_creds = GSS_C_NO_CREDENTIAL;
+    gss_buffer_desc name_token;
+    gss_name_t name;
+    int ret = AUTH_GSS_COMPLETE;
+
+    // Get credentials
+    maj_stat = gss_acquire_cred(&min_stat, GSS_C_NO_NAME, GSS_C_INDEFINITE,
+                               GSS_C_NO_OID_SET, GSS_C_INITIATE, &server_creds, NULL, NULL);
+
+    if (GSS_ERROR(maj_stat))
+    {
+       set_gss_error(maj_stat, min_stat);
+       ret = AUTH_GSS_ERROR;
+       goto end;
+    }
+
+    // Get the name
+    maj_stat = gss_inquire_cred(&min_stat, server_creds, &name, 
+                                NULL, NULL, NULL);
+
+    if (GSS_ERROR(maj_stat))
+    {
+       set_gss_error(maj_stat, min_stat);
+       ret = AUTH_GSS_ERROR;
+       goto end;
+    }
+
+    // Make the name usable
+    name_token.length = 0;
+    maj_stat = gss_display_name(&min_stat, name, &name_token, NULL);
+
+    if (GSS_ERROR(maj_stat))
+    {
+        if (name_token.value)
+            gss_release_buffer(&min_stat, &name_token);
+            gss_release_name(&min_stat, &name);
+           
+        set_gss_error(maj_stat, min_stat);
+        ret = AUTH_GSS_ERROR;
+        goto end;
+    }
+    else
+    {
+        state->username = (char *)malloc(name_token.length + 1);
+        strncpy(state->username, (char*) name_token.value, name_token.length);
+        state->username[name_token.length] = 0;
+        gss_release_buffer(&min_stat, &name_token);
+        gss_release_name(&min_stat, &name);
+    }
+
+end:
+    if (server_creds != GSS_C_NO_CREDENTIAL) {
+        maj_stat = gss_release_cred(&min_stat, &server_creds);
+    }
+
+    return ret;
+}
+
 int authenticate_gss_server_init(const char *service, gss_server_state *state)
 {
     OM_uint32 maj_stat;
Index: src/kerberosgss.h
===================================================================
--- src/kerberosgss.h	(revision 12223)
+++ src/kerberosgss.h	(working copy)
@@ -56,6 +56,7 @@
 int authenticate_gss_client_step(gss_client_state *state, const char *challenge);
 int authenticate_gss_client_unwrap(gss_client_state* state, const char* challenge);
 int authenticate_gss_client_wrap(gss_client_state* state, const char* challenge, const char* user, int protect);
+int authenticate_gss_client_inquire_cred(gss_client_state* state);
 
 int authenticate_gss_server_init(const char* service, gss_server_state* state);
 int authenticate_gss_server_clean(gss_server_state *state);
