diff -u setup.py setup.new.py
--- setup.py	2007-01-19 11:36:00.000000000 -0800
+++ setup.new.py	2007-01-19 11:34:37.000000000 -0800
@@ -18,7 +18,7 @@
     author_email = "anthony@interlink.com.au",
     url = 'http://sourceforge.net/projects/pythondirector/',
     packages = ['pydirector'],
-    scripts = ['pydir.py'],
+    scripts = ['pydir.py', 'pydir++.py'],
     classifiers = [
        'Development Status :: 5 - Production/Stable',
        'Environment :: Web Environment',
