--- dateutil/tz.py.orig	2007-05-17 15:43:48.000000000 -0400
+++ dateutil/tz.py	2007-05-17 15:45:03.000000000 -0400
@@ -709,9 +709,9 @@
         if tzid is None:
             keys = self._vtz.keys()
             if len(keys) == 0:
-                raise "no timezones defined"
+                raise ValueError, "no timezones defined"
             elif len(keys) > 1:
-                raise "more than one timezone available"
+                raise ValueError, "more than one timezone available"
             tzid = keys[0]
         return self._vtz.get(tzid)
