--- dateutil/rrule.py.orig	2007-05-17 15:43:39.000000000 -0400
+++ dateutil/rrule.py	2007-05-17 15:44:44.000000000 -0400
@@ -965,9 +965,9 @@
                                                ignoretz=ignoretz,
                                                tzinfos=tzinfos)
             except AttributeError:
-                raise "unknown parameter '%s'" % name
+                raise ValueError, "unknown parameter '%s'" % name
             except (KeyError, ValueError):
-                raise "invalid '%s': %s" % (name, value)
+                raise ValueError, "invalid '%s': %s" % (name, value)
         return rrule(dtstart=dtstart, cache=cache, **rrkwargs)
 
     def _parse_rfc(self, s,
