--- setup.py.orig	2006-02-02 16:09:41.000000000 -0800
+++ setup.py	2006-02-02 16:23:30.000000000 -0800
@@ -3,12 +3,13 @@
 from distutils.core import setup
 import glob
 import os
+import sys
+
+datadir = get_python_lib(1)[len(sys.prefix)+1:]
 
 if os.path.isfile("MANIFEST"):
     os.unlink("MANIFEST")
 
-PYTHONLIB = get_python_lib(1)
-
 setup(name="python-dateutil",
       version = "1.0",
       description = "Extensions to the standard python 2.3+ datetime module",
@@ -22,6 +23,6 @@
 datetime module, available in Python 2.3+.
 """,
       packages = ["dateutil", "dateutil.zoneinfo"],
-      data_files = [(PYTHONLIB+"/dateutil/zoneinfo", 
-                     glob.glob("dateutil/zoneinfo/zoneinfo*.tar.*"))],
+      data_files = [(os.path.join(datadir, "dateutil", "zoneinfo"),
+                     glob.glob(os.path.join("dateutil", "zoneinfo", "zoneinfo*.tar.*")))],
       )
